M2/Macaulay2/m2/code.m2

```diff
@@ -50,20 +54,36 @@ codeFunction := (f,depth) -> (
 			      comment, 
 			      if instance(val, Function) then codeFunction(val,depth+1) else net val
 			      )))))
-code = method(Dispatch => Thing)
-code Nothing := null -> null
-code Symbol := code Pseudocode := s -> getSourceLines locate s
-code Sequence := s -> (
-     f := lookup s;
-     if f =!= null then "-- code for method: " | formatDocumentTag s || code f 
-     else "-- no method function found: " | formatDocumentTag s
-     )
-code Function := f -> codeFunction(f,0)
-code List := v -> stack between_"---------------------------------" apply(v,code)
-code Command := cmd -> code cmd#0
 
+-- stores previously listed methods or hooks, to be used by (code, ZZ)
 previousMethodsFound := null
-code ZZ := i -> code previousMethodsFound#i
+
+code = method(Dispatch => Thing)
+code Nothing    := identity
+code Symbol     :=
+code Pseudocode := s -> getSourceLines locate s
+code Sequence   := s -> (
+    key := select(s, x -> not instance(x, Option));
+    -- handle strategies
+    func := if not #key === #s then (
+	opts := new OptionTable from toList select(s, x -> instance(x, Option));
+	if opts.?Strategy then (
+	    strategy := opts.Strategy;
+	    store := getHookStore(key, false);
+	    if store =!= null and store#?key
+	    and store#key.HookAlgorithms#?strategy
+	    then store#key.HookAlgorithms#strategy));
+    if func =!= null or (func = lookup key) =!= null
+    then "-- code for method: "          | formatDocumentTag key || code func
+    else "-- no method function found: " | formatDocumentTag key)
+code Function   := f -> codeFunction(f, 0)
+code Command    := C -> code C#0
+code List       := L -> stack between_"---------------------------------" apply(L, code)
+code ZZ         := i -> code previousMethodsFound#i
```

Comment:
Notice that `in hooks(ZZ, foo)`, the entries "(foo, Strategy => i)" don't include any reference to ZZ. This is because you intentionally wanted the hook to be stored in ZZ under the symbol "foo". If you want to do this, you can access the function using `ZZ#foo`. Otherwise, you should use the other way of installing a hook:
```m2
i1 : addHook((foo, ZZ),cosh)

o1 = cosh

o1 : CompiledFunction

i2 : addHook((foo, ZZ),x->x^3)

o2 = -*Function[stdio:2:20-2:24]*-

o2 : FunctionClosure

i3 : hooks (foo, ZZ)

o3 = {0 => (foo, ZZ, Strategy => 0)}
     {1 => (foo, ZZ, Strategy => 1)}

o3 : NumberedVerticalList

i4 : code oo

o4 = -- code for method: foo(ZZ)
     function cosh: source code not available
     ---------------------------------
     -- code for method: foo(ZZ)
```

Reply:
You can fix the behavior of `hooks(ZZ, foo)` to make this work somehow.

Reply:
Would you like to open an issue to fix it for the next release?

Reply:
Done

Reply:
This is potentially a reason to deprecate the `addHook(MutableHashTable, Thing, Function)` syntax for good. It's an anti-pattern.

Reply:
However, for now, this syntax is almost never used anywhere, so I don't see an urgency for either deprecating it or making hooks work the way you imagine.